### PR TITLE
Docs: world api general cleanup/overhaul

### PR DESCRIPTION
**Pull request by alwaysintreble**
## What is this fixing or adding?
Brings the style guide and world api doc inline with the more modern style guide. Makes the bullet points consistent. Small grammar improvements. Removes suggesting of generate_basic and LogicMixin. They're still mentioned, but now with more context on what they should actually be used for. Adds documentation on the fact that stage_methods exist. Recommends reST for docstrings. Adds more detail to some descriptions that seemed lacking.

## How was this tested?
👁️

### CODE REVIEW

File: docs/world api.md [hunk 1]
```diff
@@ -1,95 +1,94 @@
 # Archipelago API
 
-This document tries to explain some internals required to implement a game for
-Archipelago's generation and server. Once a seed is generated, a client or mod is 
-required to send and receive items between the game and server.
+This document tries to explain some internals required to implement a game for Archipelago's generation and server.
+Once a seed is generated, a client or mod is required to send and receive items between the game and server.
```
Comment (ReverM) [inline docs/world api.md:4]: I'm not sure I see the link between the two sentences.

File: docs/world api.md [hunk 2]
```diff
@@ -122,69 +122,75 @@ class RLWeb(WebWorld):
...
+By convention, options are defined in `options.py` and will be used when parsing the players' yaml files.
+
+Each option has its own class, which inherits from a base option type, has a docstring to describe it, and a
+`display_name` property for display on the website and in spoiler logs.
+
+The actual name as used in the yaml is defined via the field names of a `dataclass` that is assigned to the world under
+`self.options_dataclass`. By convention, the strings that define your option names should be in `snake_case`.
+
+Common option types are `Toggle`, `DefaultOnToggle`, `Choice`, and `Range`.
+For more information, see the [options api doc](options%20api.md).
```
Comment (ReverM) [inline docs/world api.md:152]: The whole section is understandable, I'm just less sure about the dataclass paragraph but I was able to understand the rest

File: docs/world api.md [hunk 3]
```diff
@@ -122,69 +122,75 @@ class RLWeb(WebWorld):
...
 
 ### World Settings
 
-Any AP installation can provide settings for a world, for example a ROM file, accessible through
-`self.settings.<setting_name>` or `cls.settings.<setting_name>` (new API)
-or `Utils.get_options()["<world>_options"]["<setting_name>"]` (deprecated).
+Settings are set by the user for the particular world, and can be used for those settings that may affect generation or
```
Comment (ReverM) [inline docs/world api.md:156]: Is "world" here refering to one player's world or a game generated with multiple worlds? What is "the particular world" here?

File: docs/world api.md [hunk 4]
```diff
@@ -259,215 +262,124 @@ class MyGameWorld(World):
 
 ### Events
 
-Events will mark some progress. You define an event location, an
-event item, strap some rules to the location (i.e. hold certain
-items) and manually place the event item at the event location.
+An Event is a special combination of a Location and an Item, with both having an `id` of `None`. These can be used to
+track certain logic interactions, with the Event Item being required for access in other locations or regions, but not
+being "real". They may also be used for making the spoiler log look nicer, i.e. by having a `"Victory"` Event Item, that
```
Comment (ReverM) [inline docs/world api.md:267]: What is meant by not "real"? Otherwise the rest of the paragraph is very clear.

File: docs/world api.md [hunk 5]
```diff
@@ -259,215 +262,124 @@ class MyGameWorld(World):
...
+    victory_loc.place_locked_item(MyGameItem("Victory", ItemClassification.progression, None, self.player))
+```
 
 ### Regions
 
-Regions are logical groups of locations that share some common access rules. If
-location logic is written from scratch, using regions greatly simplifies the
-definition and allows to somewhat easily implement things like entrance
-randomizer in logic.
+Regions are logical containers, that typically hold locations that share some common access rules. If location logic is
```
Comment (ReverM) [inline docs/world api.md:288]: Can regions hold things other than locations?

File: docs/world api.md [hunk 6]
```diff
@@ -259,215 +262,124 @@ class MyGameWorld(World):
...
-from worlds.AutoWorld import World
-from .options import MyGameOptions  # import the options dataclass
-
-
-class MyGameWorld(World):
-    # ...
-    options_dataclass = MyGameOptions  # assign the options dataclass to the world
-    options: MyGameOptions  # typing for option results
-    # ...
-```
```
Comment (ReverM) [inline docs/world api.md:470]: Is this going to be found somewhere else? I personally find examples to be really useful so removing them might not be a net positive.

**Review by ReverM — COMMENTED** [review]
I have read through both documents and left comments regarding understandability and similar. My review only looked at the content and did not consider if the markdown style was fully respected.

File: docs/world api.md [hunk 7]
```diff
@@ -259,215 +262,124 @@ class MyGameWorld(World):
...
-Events will mark some progress. You define an event location, an
-event item, strap some rules to the location (i.e. hold certain
-items) and manually place the event item at the event location.
+An Event is a special combination of a Location and an Item, with both having an `id` of `None`. These can be used to
+track certain logic interactions, with the Event Item being required for access in other locations or regions, but not
+being "real". They may also be used for making the spoiler log look nicer, i.e. by having a `"Victory"` Event Item, that
+is required to finish the game, it will be very clear where the player finishes, rather than only seeing their last
+relevant found Item. Events function just like any other Location, and can still have their own access rules, etc.
+By convention, the Event "pair" of Location and Item typically have the same name, though this is not a requirement.
+They do not need to exist in the `name_to_id` lookups, as they will be stripped at the end of generation.
```
Comment (ReverM) [inline docs/world api.md:271]:
```suggestion
They must not need to exist in the `name_to_id` lookups, as they will be stripped at the end of generation.
```
As requested by treble

Comment (black-sliver) [reply docs/world api.md:271]: > as they will be stripped at the end of generation.

This is unclear imo. I would rephrase as something like

> They must not exist in the `name_to_id` lookups, are only used during generation and will not exist in the generated output.

File: docs/world api.md (hunk 5 above)
Comment (alwaysintreble) [reply docs/world api.md:288]: yes they can also hold nothing.

File: docs/world api.md (hunk 6 above)
Comment (alwaysintreble) [reply docs/world api.md:470]: thought these were in the options api doc already. moved them now.

File: docs/world api.md [hunk 8]
```diff
@@ -202,46 +205,43 @@ class MyGameWorld(World):
...
+groups. These descriptions will show up in item-selection options in the Weighted Options page. Extra indentation and
+single newlines will be collapsed into spaces.
 
 ```python
-# Items.py
+# items.py
 
 item_descriptions = {
-    "Red Potion": "A standard health potion",
+    "Red Potion":    "A standard health potion",
```
Comment (black-sliver) [inline docs/world api.md:240]: Again, the two extra spaces do not improve readability and black would remove them.

**Review by black-sliver — COMMENTED** [review]
checked `options api.md` and `style.md`, (most of the) code in `world api.md`, only skimmed text in `world api.md`

File: docs/world api.md [hunk 9]
```diff
@@ -874,17 +777,19 @@ from test.bases import WorldTestBase
...
 
 Example `test_chest_access.py`
+
 ```python
 from . import MyGameTestBase
 
 
 class TestChestAccess(MyGameTestBase):
+
     def test_sword_chests(self) -> None:
```
[the remark this answers is not in the document]
Comment (alwaysintreble) [reply docs/world api.md:793]: i'm ok with whatever you decide, but i personally prefer to have a line above method definitions. if it were an attribute i wouldn't put the blank line but i do prefer it for a class that has a method and no defined attributes.

[the remark this answers is not in the document]
Comment (black-sliver) [reply docs/world api.md:793]: Well, imo there are cases where it hurts readability. That is
```python
class Outer:
    class Inner1:
        pass

    class Inner2:

        def method(self):
            pass

    class Inner3:
        pass
```
In BaseClasses we use the version with no blank line, so I would revert those changes.

File: docs/world api.md [hunk 10]
```diff
@@ -530,74 +439,104 @@ class MyGameWorld(World):
...
-  before, during and after the regular fill process, before `generate_output`.
-  If items need to be placed during pre_fill, these items can be determined
-  and created using `get_prefill_items`
-* `generate_output(self, output_directory: str)` that creates the output
-  files if there is output to be generated. When this is
-  called, `self.multiworld.get_locations(self.player)` has all locations for the player, with
-  attribute `item` pointing to the item.
-  `location.item.player` can be used to see if it's a local item.
+  before, during and after the regular fill process; all finishing before `generate_output`. Any items that need to be
+  placed during `pre_fill` should not exist in the itempool, and if any items that need to be filled this way, but need
```
Comment (Exempt-Medic) [inline docs/world api.md:471]: I think this is the intended reading
```suggestion
  placed during `pre_fill` should not exist in the itempool, and if there are any items that need to be filled this way, but need
```

File: docs/world api.md [hunk 11]
```diff
@@ -530,74 +439,104 @@ class MyGameWorld(World):
...
+  to be in state while you fill other items, they can be returned from `get_prefill_items`.
```
Comment (Exempt-Medic) [inline docs/world api.md:472]: Could also change "return" to a different verb like "read" or "acquired", but I think "returned" makes sense
```suggestion
  to be in state while you fill other items, they can be returned with `get_prefill_items`.
```

File: docs/world api.md [hunk 12]
```diff
@@ -730,59 +612,79 @@ def set_rules(self) -> None:
...
+# visualize_regions(self.multiworld.get_region("Menu", self.player), "my_world.puml")
+```
+
+### Custom Logic Rules
 
-#### Sample
+Custom methods can be defined for your logic rules. The access rule that ultimately gets assigned to the Location or
+Entrance should be a [`CollectionRule`](https://github.com/ArchipelagoMW/Archipelago/blob/main/worlds/generic/Rules.py#L9).
+Typically, this is done by defining a lambda expression on demand at the relevant bit, typically calling other functions,
+but this can also be achieved by defining a method with the appropriate format and assigning it directly, such as
```
Comment (Exempt-Medic) [inline docs/world api.md:636]:
```suggestion
but this can also be achieved by defining a method with the appropriate format and assigning it directly. For an example, see
```

File: docs/world api.md [hunk 13]
```diff
@@ -730,59 +612,79 @@ def set_rules(self) -> None:
...
     # ...
     def set_rules(self) -> None:
         set_rule(self.multiworld.get_location("A Door", self.player),
-                 lambda state: state.mygame_has_key(self.player))
+                 lambda state: logic.mygame_has_key(state, self.player))
+```
+
+### Logic Mixin
+
+While lambdas and events could do pretty much anything, more complex logic can be handled in logic mixins.
```
Comment (Exempt-Medic) [inline docs/world api.md:669]: Change tone to match other parts of the doc
```suggestion
While lambdas and events can do pretty much anything, more complex logic can be handled in logic mixins.
```

File: docs/world api.md [hunk 14]
```diff
@@ -730,59 +612,79 @@ def set_rules(self) -> None:
...
+
+When importing a file that defines a class that inherits from `worlds.AutoWorld.LogicMixin` the `CollectionState` class
```
Comment (Exempt-Medic) [inline docs/world api.md:671]: Missing comma
```suggestion
When importing a file that defines a class that inherits from `worlds.AutoWorld.LogicMixin`, the `CollectionState` class
```

File: docs/world api.md [hunk 15]
```diff
@@ -730,59 +612,79 @@ def set_rules(self) -> None:
...
+is automatically extended by the mixin's members. These members should be prefixed with the name of the implementing
+world since the namespace is shared with all other logic mixins.
+
+Some uses could be to add additional variables to the state object, or have a custom state machine that gets modified
```
Comment (Exempt-Medic) [inline docs/world api.md:675]:
```suggestion
Some uses could be to add additional variables to the state object, or to have a custom state machine that gets modified
```

File: docs/world api.md [hunk 16]
```diff
@@ -824,20 +726,19 @@ def generate_output(self, output_directory: str) -> None:
...
 
 If the game client needs to know information about the generated seed, a preferred method of transferring the data
-is through the slot data. This can be filled from the `fill_slot_data` method of your world by returning a `Dict[str, Any]`,
-but should be limited to data that is absolutely necessary to not waste resources. Slot data is sent to your client once
-it has successfully [connected](network%20protocol.md#connected).
-If you need to know information about locations in your world, instead
-of propagating the slot data, it is preferable to use [LocationScouts](network%20protocol.md#locationscouts) since that
-data already exists on the server. The most common usage of slot data is to send option results that the client needs
-to be aware of.
+is through the slot data. This can be filled from the `fill_slot_data` method of your world by returning
```
Comment (Exempt-Medic) [inline docs/world api.md:729]: "from" is a bit of a weird preposition, could also use "through" or "using" instead of "with"
```suggestion
is through the slot data. This can be filled with the `fill_slot_data` method of your world by returning
```

File: docs/world api.md [hunk 17]
```diff
@@ -824,20 +726,19 @@ def generate_output(self, output_directory: str) -> None:
...
+a `Dict[str, Any]`, but should be limited to data that is absolutely necessary to not waste resources. Slot data is
```
Comment (Exempt-Medic) [inline docs/world api.md:730]: Removes a potential ambiguity by reordering the phrases
```suggestion
a `Dict[str, Any]`, but, to not waste resources, this should be limited to data that is absolutely necessary. Slot data is
```

File: docs/world api.md [hunk 18]
```diff
@@ -824,20 +726,19 @@ def generate_output(self, output_directory: str) -> None:
...
+sent to your client once it has successfully [connected](network%20protocol.md#connected).
+If you need to know information about locations in your world; instead of propagating the slot data, it is preferable
```
Comment (Exempt-Medic) [inline docs/world api.md:732]:
```suggestion
If you need to know information about locations in your world, instead of propagating the slot data, it is preferable
```

File: docs/world api.md [hunk 19]
```diff
@@ -824,20 +726,19 @@ def generate_output(self, output_directory: str) -> None:
...
+to use [LocationScouts](network%20protocol.md#locationscouts), since that data already exists on the server. The most
+common usage of slot data is to send option results that the client needs to be aware of.
```
Comment (Exempt-Medic) [inline docs/world api.md:734]:
```suggestion
common usage of slot data is sending option results that the client needs to be aware of.
```

File: docs/world api.md [hunk 20]
```diff
@@ -847,12 +748,14 @@ Each world implementation should have a tutorial and a game info page. These are
...
+
 The game info page is for a short breakdown of what your game is and how it works in Archipelago. Any additional
 information that may be useful to the player when learning your randomizer should also go here. The file name format
 is `<language key>_<game name>.md`. While you can write these docs for multiple languages, currently only the english
 version is displayed on the website.
 
 #### Tutorials
+
 Your game can have as many tutorials in as many languages as you like, with each one having a relevant `Tutorial`
 defined in the `WebWorld`. The file name you use aren't particularly important, but it should be descriptive of what
```
Comment (Exempt-Medic) [inline docs/world api.md:760]:
```suggestion
defined in the `WebWorld`. The file name you use isn't particularly important, but it should be descriptive of what
```

File: docs/world api.md [hunk 21]
```diff
@@ -847,12 +748,14 @@ Each world implementation should have a tutorial and a game info page. These are
...
 the tutorial is covering, and the name of the file must match the relative URL provided in the `Tutorial`. Currently,
```
Comment (Exempt-Medic) [inline docs/world api.md:761]:
```suggestion
the tutorial covers, and the name of the file must match the relative URL provided in the `Tutorial`. Currently,
```

File: docs/world api.md [hunk 22]
```diff
@@ -874,17 +777,19 @@ from test.bases import WorldTestBase
 
 
 class MyGameTestBase(WorldTestBase):
-  game = "My Game"
+    game = "My Game"
 ```
 
 Next using the rules defined in the above `set_rules` we can test that the chests have the correct access rules.
```
Comment (Exempt-Medic) [inline docs/world api.md:783]:
```suggestion
Next, using the rules defined in the above `set_rules`, we can test that the chests have the correct access rules.
```

File: docs/world api.md [hunk 23]
```diff
@@ -874,17 +777,19 @@ from test.bases import WorldTestBase
...
+
 ```python
 from . import MyGameTestBase
 
 
 class TestChestAccess(MyGameTestBase):
+
     def test_sword_chests(self) -> None:
         """Test locations that require a sword"""
         locations = ["Chest1", "Chest2"]
```
Comment (Exempt-Medic) [inline docs/world api.md:795]: It won't let me comment on line 808 (the last line), but it's missing a comma after "For more information on tests"

File: docs/options api.md [hunk 24]
```diff
@@ -27,28 +27,49 @@ Choice, and defining `alias_true = option_full`.
 - All options support `random` as a generic option. `random` chooses from any of the available values for that option,
 and is reserved by AP. You can set this as your default value, but you cannot define your own `option_random`.
 
-As an example, suppose we want an option that lets the user start their game with a sword in their inventory. Let's
-create our option class (with a docstring), give it a `display_name`, and add it to our game's options dataclass:
+As an example, suppose we want an option that lets the user start their game with a sword in their inventory, an option
+to let the player choose the difficulty, and one to choose how much health the final boss has. Let's create our option
```
Comment (Exempt-Medic) [inline docs/options api.md:31]: This is probably the best option for making the list consistent in wording
```suggestion
to let the player choose the difficulty, and an option to choose how much health the final boss has. Let's create our option
```

**Review by Exempt-Medic — CHANGES_REQUESTED** [review]
Read through and made some grammar/wording/style suggestions

File: docs/world api.md (hunk 11 above)
Comment (alwaysintreble) [reply docs/world api.md:472]: not exactly. `get_prefill_items` returns a `List[Item]` and it gets called specifically by `get_all_state`, not the world itself. there probably could be better wording but the suggestion here is slightly incorrect.

File: docs/world api.md [hunk 25]
```diff
@@ -1,95 +1,95 @@
...
-
+Clients that connect to the server to sync items can be in any language that allows using WebSockets.
 
 ## Coding style
 
-AP follows [style.md](https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/style.md).
-When in doubt use an IDE with coding style linter, for example PyCharm Community Edition.
-
+AP follows a [style guide](https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/style.md).
+When in doubt, use an IDE with coding style linter, for example PyCharm Community Edition.
```
Comment (Exempt-Medic) [inline docs/world api.md:22]: The "a" should at least be here, less sure about "coding style" vs "code-style", but I think the latter is correct
```suggestion
When in doubt, use an IDE with a code-style linter, for example PyCharm Community Edition.
```

File: docs/world api.md [hunk 26]
```diff
@@ -1,95 +1,95 @@
...
-A `WebWorld` class contains specific attributes and methods that can be modified
-for your world specifically on the webhost:
+A `WebWorld` class contains specific attributes and methods that can be modified for your world specifically on the
+webhost:
 
-`settings_page`, which can be changed to a link instead of an AP generated settings page.
+* `options_page` can be changed to a link instead of an AP-generated options page.
 
-`theme` to be used for your game specific AP pages. Available themes:
+* `theme` to be used for your game-specific AP pages. Available themes:
```
Comment (Exempt-Medic) [inline docs/world api.md:59]: I think something went wrong with the display here, it doesn't look right when selecting "View file"

**Review by Exempt-Medic — CHANGES_REQUESTED** [review]

**Review by Exempt-Medic — APPROVED** [review]

**Review by ReverM — APPROVED** [review]
I've reread the changes and they are more understandable now
LGTM

**Review by black-sliver — APPROVED** [review]
Thanks!
